File: JavaScript/mergeSort.js
function mergeSort(arr) {
   if (arr.length <= 1) return arr;
//    get the midpoint of the array
   const mid = Math.floor(arr.length / 2);
//    get both the left and right halves of the array
   const left = arr.slice(0, mid);
   const right = arr.slice(mid);
//    
   return merge(mergeSort(left), mergeSort(right));
}




// helper function 'merge' that takes in two arrays, the left and right halves
function merge(left, right) {
// instantiate a dummy array
   const result = [];
// while the left and right arrays still have elements in them:
   while (left.length && right.length) {
       if (left[0] < right[0]) {
          // if the nth index of the lArray is smaller than that of rArray:
            // push the first element of lArray to 'result', then shift it out
         result.push(left.shift());
      } else {
        // vice versa if vice versa
         result.push(right.shift());
      }
   }
    
   return [...result, ...left, ...right];
}

File: Ruby/merge_sort.rb
def merge_sort(arr)
    return arr if arr.length<=1
    mid = (arr.length/2).floor()
    left = arr[0..mid]
    right = arr[mid..-1]
    p "almost"
    merge(merge_sort(arr[0...mid]), merge_sort(arr[mid..-1]))
end

def merge(left, right)
    p "in merge"
    res = []
    until left.length==0 || right.length==0 do
        res << (left.first <= right.first ? left.shift : right.shift)
    end
    res + left + right
end


p merge_sort([1,9,3,8,4,72,16])